### PR TITLE
fix: Update company picker keyboard navigation

### LINKED ISSUE #909: Update company picker to navigate with arrow keys and add a border
We miss two things:
- I can use keyboard to switch from list items to create button with down/up arrow
- a light border above the create button

Desired:

Current:
image

### PR DESCRIPTION
**Pull request by G3root**
fixes  #909

[Screencast 2023-09-17 10:06:43.webm](https://github.com/twentyhq/twenty/assets/84864519/953e3819-a8d6-44f9-8147-15e8fb0598f2)

### CODE REVIEW

**Review by lucasbordeau — CHANGES_REQUESTED** [review]

**Comment by G3root** [comment]
@lucasbordeau pushed some fixes for the requested changes

**Review by lucasbordeau — CHANGES_REQUESTED** [review]
It won't work with this current implementation. You need to think with Recoil.

Things to avoid in your implementation : 
- Direct DOM manipulation with classes
- Drilling down refs
- Trying to solve this problem without using Recoil

Create as many atoms as you need so that each component is independent from the other and only uses atoms to compute whether it is hovered or not.

Hint of a solution with Recoil : 
- Create an atom hoverableIds that olds an array of string
- In SingleEntitySelect, or another place where you get the entities, fill this atom with the entity ids, and if there is an onCreateHandler passed, add a last "create-button" id in this array
- Create a currentlyHoveredId atom
- In SingleEntitySelect, in up and down hotkey callback, instead of incrementing a hoverIndex, go to the next id in the array and set currentlyHoveredId to this id, if it's the last go to "create-button" id
- Create a new component PickerCreateButton, that wraps MenuItem, and that calls currentlyHoveredId to know if it's being hovered. This component should expose and onClick event that takes the handleCreate callback

It's a bit more work but you'll get how we use Recoil in our codebase. Keep it up.

**Comment by G3root** [comment]
hi @lucasbordeau pushed some fixes for the requested changes. This is my first experience with Recoil, so I'm not entirely familiar with how Recoil is utilized here . happy to have some feedback on this 🙏🏼

File: front/src/modules/ui/input/relation-picker/components/SingleEntitySelectBase.tsx [hunk 1]
```diff
@@ -38,26 +41,39 @@ export const SingleEntitySelectBase = <
...
 
   useScopedHotkeys(
     Key.Enter,
     () => {
-      onEntitySelected(entitiesInDropdown[hoveredIndex]);
+      if (onCreateHandler && hoveredId === CreateButtonId) {
+        onCreateHandler();
+      } else {
+        const index = entitiesInDropdown.findIndex(
+          (val) => val.id === hoveredId,
```
Comment (lucasbordeau) [inline front/src/modules/ui/input/relation-picker/components/SingleEntitySelectBase.tsx:68]: Always try to keep a clear naming even for small parts of the code, here it should be at least `entity`

File: front/src/modules/ui/input/relation-picker/hooks/useEntitySelectScroll.ts [hunk 2]
```diff
@@ -4,24 +4,30 @@ import { Key } from 'ts-key-enum';
...
+import { relationPickerHoverIdScopedState } from '../states/relationPickerHoverIdScopedState';
 import { RelationPickerHotkeyScope } from '../types/RelationPickerHotkeyScope';
 
-export const useEntitySelectScroll = <
-  CustomEntityForSelect extends EntityForSelect,
->({
+export const useEntitySelectScroll = ({
   containerRef,
-  entities,
+  hoverIds,
```
Comment (lucasbordeau) [inline front/src/modules/ui/input/relation-picker/hooks/useEntitySelectScroll.ts:14]: `hoveredIds`

File: front/src/modules/ui/input/relation-picker/states/relationPickerHoverIdScopedState.ts [hunk 3]
```diff
@@ -0,0 +1,6 @@
+import { atomFamily } from 'recoil';
+
+export const relationPickerHoverIdScopedState = atomFamily<string, string>({
```
Comment (lucasbordeau) [inline front/src/modules/ui/input/relation-picker/states/relationPickerHoverIdScopedState.ts:3]: `relationPickerHoveredIdScopedState`

File: front/src/modules/ui/input/relation-picker/hooks/useEntitySelectScroll.ts [hunk 4]
```diff
@@ -60,38 +64,36 @@ export const useEntitySelectScroll = <
...
     Key.ArrowDown,
     () => {
-      setRelationPickerHoverIndex((prevSelectedIndex) =>
-        Math.min(prevSelectedIndex + 1, (entities?.length ?? 0) - 1),
-      );
-
-      const currentHoveredRef = containerRef.current?.children[
-        relationPickerHoverIndex
-      ] as HTMLElement;
+      const nextId = hoverIds[nextIndex];
```
Comment (lucasbordeau) [inline front/src/modules/ui/input/relation-picker/hooks/useEntitySelectScroll.ts:73]: Please always try to keep a maximum of context in your variable naming, avoid generic names and also shortened names.

Here it should be at something like `nextHoverableId` to tell that we're using this variable for guessing what will be the next item in the list.

File: front/src/modules/ui/input/relation-picker/states/recoil-scope-contexts/RelationPickerRecoilScopeContext.ts [hunk 5]
```diff
@@ -0,0 +1,5 @@
+import { createContext } from 'react';
+
+export const RelationPickerRecoilScopeContext = createContext<string | null>(
+  'relationship-context',
```
Comment (lucasbordeau) [inline front/src/modules/ui/input/relation-picker/states/recoil-scope-contexts/RelationPickerRecoilScopeContext.ts:4]: `relation-picker-context`

File: front/src/modules/ui/input/relation-picker/hooks/useEntitySelectScroll.ts [hunk 6]
```diff
@@ -4,24 +4,30 @@ import { Key } from 'ts-key-enum';
...
   containerRef: React.RefObject<HTMLDivElement>;
 }) => {
-  const [relationPickerHoverIndex, setRelationPickerHoverIndex] =
-    useRecoilScopedState(relationPickerHoverIndexScopedState);
+  const [relationPickerHoverId, setRelationPickerHoverId] =
+    useRecoilScopedState(
+      relationPickerHoverIdScopedState,
+      RelationPickerRecoilScopeContext,
+    );
+  const getIndex = hoverIds.findIndex((val) => val === relationPickerHoverId);
```
Comment (lucasbordeau) [inline front/src/modules/ui/input/relation-picker/hooks/useEntitySelectScroll.ts:24]: You can put this in a selector, see : https://recoiljs.org/docs/api-reference/utils/selectorFamily And you have an example here : `numberOfTableColumnsScopedSelector`

Comment (lucasbordeau) [reply front/src/modules/ui/input/relation-picker/hooks/useEntitySelectScroll.ts:24]: If it's too complicated with a selector, try to create a util function at least to abstract this away.

**Review by lucasbordeau — CHANGES_REQUESTED** [review]
Great work, I left a few comments for clean code

**Comment by FelixMalfait** [comment]
This is for #909 right? Linking the issue to make sure we close it after this PR is merged

File: front/src/modules/ui/input/relation-picker/components/CreateNewButton.tsx [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+
+const StyledCreateNewButton = styled(MenuItem)<{ hovered: boolean }>`
+  ${({ hovered, theme }) =>
+    hovered &&
+    css`
+      background: ${theme.background.transparent.light};
+    `}
+`;
+
+export const CreateNewButton = StyledCreateNewButton;
```
Comment (charlesBochet) [inline front/src/modules/ui/input/relation-picker/components/CreateNewButton.tsx:14]: @G3root exporting it as CreateNewButton to hide the fact that this is only a Styled component. The outside world shouldn't care about it!

File: front/src/modules/ui/input/relation-picker/components/SingleEntitySelect.tsx [hunk 8]
```diff
@@ -71,15 +69,12 @@ export const SingleEntitySelect = <
         autoFocus
       />
       <StyledDropdownMenuSeparator />
-      <SingleEntitySelectBase {...props} onCancel={onCancel} />
-      {showCreateButton && (
```
Comment (charlesBochet) [inline front/src/modules/ui/input/relation-picker/components/SingleEntitySelect.tsx:75]: moving it withing SingleSelectBase to simplify the code

File: front/src/modules/ui/input/relation-picker/components/SingleEntitySelectBase.tsx [hunk 9]
```diff
@@ -38,26 +45,40 @@ export const SingleEntitySelectBase = <
...
   const entitiesInDropdown = [selectedEntity, ...entitiesToSelect].filter(
     (entity): entity is CustomEntityForSelect =>
       assertNotNull(entity) && isNonEmptyString(entity.name.trim()),
   );
 
-  const { hoveredIndex, resetScroll } = useEntitySelectScroll({
-    entities: entitiesInDropdown,
+  const { preselectedOptionId, resetScroll } = useEntitySelectScroll({
+    selectableOptionIds: [
+      EmptyButtonId,
```
Comment (charlesBochet) [inline front/src/modules/ui/input/relation-picker/components/SingleEntitySelectBase.tsx:59]: adding EmptyButton possibility

File: front/src/modules/ui/input/relation-picker/hooks/useEntitySelectScroll.ts [hunk 10]
```diff
@@ -4,24 +4,32 @@ import { Key } from 'ts-key-enum';
...
 import { RelationPickerHotkeyScope } from '../types/RelationPickerHotkeyScope';
+import { getHoveredIdIndex } from '../utils/getHoveredIdIndex';
 
-export const useEntitySelectScroll = <
-  CustomEntityForSelect extends EntityForSelect,
->({
+export const useEntitySelectScroll = ({
   containerRef,
-  entities,
+  selectableOptionIds,
```
Comment (charlesBochet) [inline front/src/modules/ui/input/relation-picker/hooks/useEntitySelectScroll.ts:15]: renaming hoveredId to selectableOptionIds

File: front/src/modules/ui/input/relation-picker/hooks/useEntitySelectScroll.ts [hunk 11]
```diff
@@ -60,38 +67,40 @@ export const useEntitySelectScroll = <
...
       }
     },
     RelationPickerHotkeyScope.RelationPicker,
-    [setRelationPickerHoverIndex, entities],
+    [selectableOptionIds],
   );
 
   return {
-    hoveredIndex: relationPickerHoverIndex,
+    preselectedOptionId: relationPickerHoveredId,
```
Comment (charlesBochet) [inline front/src/modules/ui/input/relation-picker/hooks/useEntitySelectScroll.ts]: and this one to preselectedOptionId which is more accurate!

**Review by charlesBochet — COMMENTED** [review]
Made a few modifications @G3root @lucasbordeau

**Pull request merged**